### ISSUE TITLE
[unbound-method] inconsistent result

### ISSUE THREAD

**Issue by TrejGun** [typescript-eslint/typescript-eslint#2245]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/unbound-method": "error"
  }
}
```

```TS
import React, {Fragment, FC, useEffect} from "react";
import {useSnackbar} from "notistack";
import {useIntl} from "react-intl";


export const Test: FC = () => {
  const {enqueueSnackbar} = useSnackbar();
  const {formatMessage} = useIntl();

  useEffect(() => {
    enqueueSnackbar(formatMessage({id: "snackbar.error"}), {variant: "error"});
  }, []);

  return <Fragment>test</Fragment>;
};
```

**Expected Result**
I expect no error

**Actual Result**
While both libraries returns unbound methods only `formatMessage` is reported

**Additional Info**
I really appreciate your efforts but recently this plugin got too many new unconfigurable rules. Please stop pushing us towards what you are thinking is good and let us code business logic instead of searching for workarounds over linter's strictness

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.4.0` |
| `@typescript-eslint/parser`        | `3.4.0` |
| `TypeScript`                       | `3.9.5` |
| `ESLint`                           | `7.3.1` |
| `node`                             | `12.16.3` |
| `npm`                              | `6.14.5` |

**Comment by TrejGun** [comment]
<img width="725" alt="Screen Shot 2020-06-24 at 3 50 54 PM" src="https://user-images.githubusercontent.com/185285/85519390-02d09480-b634-11ea-88e5-b3dab6132b95.png">

**Comment by bradzacher** [comment]
Could you please provide a more isolated repro?
You have a lot of dependencies here - including the types of `useSnackbar` and `useIntl` would help in diagnosing the issue.

**Comment by TrejGun** [comment]
More isolated example? But I have this error only here, once i see it elsewhere i add a comment to this issue
The best thing I can do is a git repo with this code, not sure if it is very helpful

**Comment by bradzacher** [comment]
The problem is that in order to reproduce this, I have to determine your exact dependencies, install them, then copy paste the code.

An isolated repro repo would help, as it would mean I can reproduce the issue by just cloning the repo.
Alternately, if you inspect the types for the mentioned functions, and include the reported types, then I would likely be able to diagnose this just within the issue itself.

**Comment by TrejGun** [comment]
check it out
https://github.com/TrejGun/unbound-method

**Comment by bradzacher** [comment]
This due to differences in how the two types are defined.

`enqueueSnackbar` is defined as a property that is a function type
https://github.com/iamhosseindhv/notistack/blob/34a7010c39035bf18bdc6317d6a85e067f7b4ccb/src/index.d.ts#L263

`formatMessage` is defined as a method definition
https://github.com/formatjs/formatjs/blob/18d5dc334d7c88ad75817f42961c4acb90dd2e88/packages/react-intl/src/types.ts#L119-L129

These are currently treated as different cases by the rule - the former is just treated as a property, the later is treated as a method.
This is why only `useIntl` reports an error.

This aligns with how we currently treat class properties vs class methods.

I think there is a bug here though in that properties on interfaces and object types with a function type should probably be treated the same method definitions.

Note - in your codebase, it's likely that you can just turn off `unbound-method`. In the functional react world, I don't think there'll be a huge amount of value in it, as you likely will never have a function that actually cares about the `this` context.

**Comment by JoshuaKGoldberg** [comment]
> I think there is a bug here though in that properties on interfaces and object types with a function type should probably be treated the same method definitions.

@bradzacher sorry, I'm not seeing the bug - could you elaborate please?

**Comment by JoshuaKGoldberg** [comment]
Closing as aged away. If anybody has an isolated repro that shows an issue with the unbound-method lint rule, please do post a new issue.
